From cf0c09caeb9e6464017514e8859ac74991714de8 Mon Sep 17 00:00:00 2001
From: Mika Laitio <lamikr@gmail.com>
Date: Sat, 31 Aug 2024 23:28:55 -0700
Subject: [PATCH 09/40] add HSA_RUNTIME_64_LIB

- cmake library search works differently on ubuntu 24.04
  compared to other distros.and fails to
  build pytorch because it not finding the hsa-runtime64.
  It seems that cmake in ubuntu can not search recursively
  subdirectories, and to get get pytorch building on
  on ubuntu we need to introduce the
  HSA_RUNTIME_LIBS and add that explicitly to search path
- disable building of pytorch tests
  which will require special handling
  to link agains hsa-runtime64. On other distros
  these tests are be disabled by default...

Signed-off-by: Mika Laitio <lamikr@gmail.com>
---
 build_rocm.sh                   | 2 +-
 cmake/public/LoadHIP.cmake      | 3 +++
 torch/lib/libshm/CMakeLists.txt | 1 +
 3 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/build_rocm.sh b/build_rocm.sh
index 5dedc9b9f8c..8215fbac041 100755
--- a/build_rocm.sh
+++ b/build_rocm.sh
@@ -20,4 +20,4 @@ unset LDFLAGS
 unset CFLAGS
 unset CPPFLAGS
 unset PKG_CONFIG_PATH
-USE_FLASH_ATTENTION=ON AOTRITON_INSTALLED_PREFIX=${install_dir_prefix_rocm} ROCM_PATH=${install_dir_prefix_rocm} ROCM_SOURCE_DIR=${install_dir_prefix_rocm} CMAKE_CXX_FLAGS="$CMAKE_CXX_FLAGS -Wno-error=maybe-uninitialized" CMAKE_PREFIX_PATH="${install_dir_prefix_rocm};${install_dir_prefix_rocm}/lib64/cmake;${install_dir_prefix_rocm}/lib/cmake;${install_dir_prefix_rocm}/lib64;${install_dir_prefix_rocm}/lib" ROCM_VERSION=${rocm_version_str} HIP_ROOT_DIR=${install_dir_prefix_rocm} USE_ROCM=1 PYTORCH_BUILD_VERSION="$(git describe --tags --abbrev=0 | sed 's/^v//')" PYTORCH_BUILD_NUMBER=1 python setup.py bdist_wheel
+BUILD_TEST=0 USE_FLASH_ATTENTION=ON AOTRITON_INSTALLED_PREFIX=${install_dir_prefix_rocm} ROCM_PATH=${install_dir_prefix_rocm} ROCM_SOURCE_DIR=${install_dir_prefix_rocm} CMAKE_CXX_FLAGS="$CMAKE_CXX_FLAGS -Wno-error=maybe-uninitialized" CMAKE_PREFIX_PATH="${install_dir_prefix_rocm};${install_dir_prefix_rocm}/lib64/cmake;${install_dir_prefix_rocm}/lib/cmake;${install_dir_prefix_rocm}/lib64;${install_dir_prefix_rocm}/lib" ROCM_VERSION=${rocm_version_str} HIP_ROOT_DIR=${install_dir_prefix_rocm} USE_ROCM=1 PYTORCH_BUILD_VERSION="$(git describe --tags --abbrev=0 | sed 's/^v//')" PYTORCH_BUILD_NUMBER=1 python setup.py bdist_wheel
diff --git a/cmake/public/LoadHIP.cmake b/cmake/public/LoadHIP.cmake
index 70fe47a9e64..a58db436142 100644
--- a/cmake/public/LoadHIP.cmake
+++ b/cmake/public/LoadHIP.cmake
@@ -160,6 +160,7 @@ if(HIP_FOUND)
 
   find_package_and_print_version(hip REQUIRED)
   find_package_and_print_version(hsa-runtime64 REQUIRED)
+  #message(FATAL_ERROR "hsa-runtime64: ${hsa-runtime64}")
   find_package_and_print_version(amd_comgr REQUIRED)
   find_package_and_print_version(rocrand REQUIRED)
   find_package_and_print_version(hiprand REQUIRED)
@@ -194,6 +195,8 @@ if(HIP_FOUND)
   find_library(ROCM_HIPRTC_LIB hiprtc HINTS ${ROCM_PATH}/lib ${ROCM_PATH}/lib64)
   # roctx is part of roctracer
   find_library(ROCM_ROCTX_LIB roctx64 HINTS ${ROCM_PATH}/lib ${ROCM_PATH}/lib64)
+  # linking does not work otherwise on ubuntu 24.04
+  find_library(HSA_RUNTIME64_LIB hsa-runtime64 HINTS ${ROCM_PATH}/lib ${ROCM_PATH}/lib64)
 
   # check whether HIP declares new types
   set(file "${PROJECT_BINARY_DIR}/hip_new_types.cc")
diff --git a/torch/lib/libshm/CMakeLists.txt b/torch/lib/libshm/CMakeLists.txt
index 8a7329ddab7..fba1921bbce 100644
--- a/torch/lib/libshm/CMakeLists.txt
+++ b/torch/lib/libshm/CMakeLists.txt
@@ -65,6 +65,7 @@ if(BUILD_LIBTORCHLESS)
 else()
   # we need to link directly to c10 here otherwise we miss symbols
   target_link_libraries(torch_shm_manager PRIVATE shm c10)
+  target_link_libraries(torch_shm_manager PRIVATE ${HSA_RUNTIME64_LIB})
 endif()
 set_target_properties(torch_shm_manager PROPERTIES
   INSTALL_RPATH "${_rpath_portable_origin}/../lib")
-- 
2.43.0

